lib/src/main/java/graphql/nadel/validation/util/NadelGetReachableTypes.kt

```diff
@@ -190,159 +176,68 @@ private class NadelReferencedTypeVisitor(
     ): Boolean {
-        onTypeReferenced(node.type.unwrapAll().name)
         return true
     }
 
     override fun visitGraphQLInputObjectType(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLInputObjectType,
+        element: NadelSchemaTraverserElement.InputObjectType,
     ): Boolean {
-        visitTypeGuard(node) { return false }
+        visitTypeGuard(element) { return false }
+        val node = element.node
         onTypeReferenced(node.name)
         return true
     }
 
-    override fun visitGraphQLList(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLList,
-    ): Boolean {
-        onTypeReferenced(node.unwrapAll().name)
-        return true
-    }
-
-    override fun visitGraphQLNonNull(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLNonNull,
-    ): Boolean {
-        onTypeReferenced(node.unwrapAll().name)
-        return true
-    }
-
     override fun visitGraphQLObjectType(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLObjectType,
+        element: NadelSchemaTraverserElement.ObjectType,
     ): Boolean {
-        visitTypeGuard(node) { return false }
-
-        // Don't look at union members defined by external services
-        if (parent is GraphQLUnionType && isUnionMemberExempt(service, parent, node)) {
-            return false
-        }
-
+        visitTypeGuard(element) { return false }
+        val node = element.node
         onTypeReferenced(node.name)
         return true
     }
 
     override fun visitGraphQLScalarType(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLScalarType,
-    ): Boolean {
-        visitTypeGuard(node) { return false }
-        onTypeReferenced(node.name)
-        return true
-    }
-
-    override fun visitGraphQLTypeReference(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLTypeReference,
-    ): Boolean {
-        onTypeReferenced(node.name)
-        return true
-    }
-
-    override fun visitGraphQLModifiedType(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLModifiedType,
-    ): Boolean {
-        onTypeReferenced(node.unwrapAll().name)
-        return true
-    }
-
-    override fun visitGraphQLCompositeType(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLCompositeType,
+        element: NadelSchemaTraverserElement.ScalarType,
     ): Boolean {
+        visitTypeGuard(element) { return false }
+        val node = element.node
         onTypeReferenced(node.name)
         return true
     }
 
     override fun visitGraphQLDirective(
-        parent: GraphQLNamedSchemaElement?,
-        node: GraphQLDirective,
+        element: NadelSchemaTraverserElement.Directive,
     ): Boolean {
         // We don't care about directives, not a Nadel runtime concern
         // GraphQL Java will do validation on them for us
         return false
     }
 
-    override fun visitGraphQLDirectiveContainer(
```

Comment:
These were never actually invoked, the fast schema traverser skips these as we don't need them.